crates/storage/src/collections/mapping/mod.rs

```diff
@@ -0,0 +1,123 @@
+// Copyright 2018-2021 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A simple mapping to contract storage.
+
+use crate::traits::{
+    push_spread_root,
+    pull_spread_root,
+    clear_spread_root,
+    ExtKeyPtr,
+    KeyPtr,
+    SpreadLayout,
+};
+use ink_env::hash::{
+    Blake2x256,
+    HashOutput,
+};
+use ink_primitives::Key;
+
+use core::marker::PhantomData;
+
+/// A mapping of key-value pairs directly into contract storage.
+///
+/// If a key does not exist the `Default` value for the `value` will be returned.
+pub struct Mapping<K, V> {
+    key: Key,
+    _phantom_mapping: (PhantomData<K>, PhantomData<V>),
+}
+
+impl<K, V> Mapping<K, V>
+where
+    K: scale::Encode,
+    V: scale::Encode + scale::Decode + Default,
+{
+    /// Creates a new empty `Mapping`.
+    ///
+    /// Not sure how this should be exposed/initialize irl.
+    pub fn new(key: Key) -> Self {
+        Self {
+            key,
+            _phantom_mapping: (Default::default(), Default::default()),
+        }
+    }
+
+    /// Insert the given `value` to the contract storage.
+    pub fn insert(&mut self, key: K, value: V) {
+        ink_env::set_contract_storage(&self.key(key), &value);
```

Comment:
We should use the `packed_{push, pull}` functions here instead of these super low level ones.